javascript/ql/src/Security/CWE-116/InvalidEntityTranscoding.ql

```diff
@@ -0,0 +1,135 @@
+/**
+ * @name Invalid entity transcoding
+ * @description When encoding HTML special characters as HTML entities, the ampersand has to
+ *              be encoded first to avoid double-encoding, and conversely it has to be decoded
+ *              last to avoid double-decoding.
+ * @kind problem
+ * @problem.severity warning
+ * @precision high
+ * @id js/invalid-entity-transcoding
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-116
+ */
+
+import javascript
+
+/**
+ * Holds if `rl` is a simple constant, which is bound to the result of the predicate.
+ *
+ * For example, `/a/g` has string value `"a"` and `/abc/` has string value `"abc"`,
+ * while `/ab?/` and `/a(?=b)/` do not have a string value. Flags are ignored.
+ *
+ * Being a simple constant is not the same as matching exactly one string:
+ * `/a(?=b)/` matches exactly one string, but is not a simple constant; `/a/i` matches
+ * two strings, but is a simple constant.
+ *
+ * Note the somewhat subtle use of monotonic aggregate semantics, which makes the
+ * `strictconcat` fail if one of the children of the root is not a constant (legacy
+ * semantics would simply skip such children).
+ */
+language[monotonicAggregates]
+string getStringValue(RegExpLiteral rl) {
+  exists (RegExpTerm root | root = rl.getRoot() |
+    result = root.(RegExpConstant).getValue()
+    or
+    result = strictconcat(RegExpTerm ch, int i |
+      ch = root.(RegExpSequence).getChild(i) |
+      ch.(RegExpConstant).getValue() order by i
+    )
+  )
+}
+
+/**
+ * Gets a predecessor of `nd` that is not an SSA phi node.
+ */
+DataFlow::Node getASimplePredecessor(DataFlow::Node nd) {
+  result = nd.getAPredecessor() and
+  not nd.(DataFlow::SsaDefinitionNode).getSsaVariable().getDefinition() instanceof SsaPhiNode
+}
+
+/**
+ * A call to `String.prototype.replace` that replaces all instances of a pattern.
+ */
+class Replacement extends DataFlow::Node {
+  RegExpLiteral pattern;
```

Comment:
I take it string patterns here will always be reported by IncompleteSanitization, and that's why we don't include them here? Maybe worth leaving a comment about this.

Reply:
Correct; I'll add a note.